<a href="https://colab.research.google.com/github/mesushan/ANN-for-fashion-MNIST-dataset/blob/master/Building_an_Artificial_Neural_Network_in_TensorFlow_2_0_for_fashion_MNIST_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data Source: [Kaggle](https://www.kaggle.com/zalando-research/fashionmnist)

## Step 1: Installing dependencies and setting up a GPU environment

In [0]:
!pip install tensorflow-gpu==2.1-rc0

## Step 2: Importing the libraries and dataset

In [0]:
import numpy as np
import datetime
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist

In [0]:
tf.__version__

'2.1.0-rc0'

## Step 3: Data Preprocessing

### Loading the dataset

In [0]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

### Normalizing the images

We divide each pixel of the image in the training and test sets by the maximum number of pixels (255).

In this way each pixel will be in the range [0, 1]. By normalizing images we make sure that our model (ANN) trains faster.

In [0]:
X_train = X_train / 255.0

In [0]:
X_test = X_test / 255.0

### Reshaping the dataset

Since we are building a fully connected network, we reshape the training set and the test set to be into the vector format.

In [0]:
# Since each image's dimension is 28x28, we reshape the full dataset to [-1 (all elements), height * width]
X_train = X_train.reshape(-1, 28*28)

In [0]:
X_train.shape

(60000, 784)

In [0]:
X_test.shape

(10000, 28, 28)

In [0]:
# We reshape the test set the same way as train set
X_test = X_test.reshape(-1, 28*28)

In [0]:
X_test.shape

(10000, 784)

## Step 4: Building an Artificial Neural Network

### Defining the model

We simply define an object of the Sequential model.

In [0]:
model = tf.keras.models.Sequential()

### Adding a first fully-connected hidden layer

Layer hyper-parameters to be used:
- number of units/neurons: 128
- activation function: ReLU
- input_shape: (784, )

In [0]:
model.add(tf.keras.layers.Dense(units=128, activation='relu', input_shape=(784, )))

### Adding a second layer with Dropout

Dropout is being used as it is a Regularization technique where we randomly set neurons in a layer to zero. That way while training those neurons won't be updated. Because some percentage of neurons won't be updated the whole training process is long and we have less chance for overfitting.

In [0]:
model.add(tf.keras.layers.Dropout(0.2))

### Adding the output layer

- units: number of classes (10 in the Fashion MNIST dataset)
- activation: softmax

In [0]:
model.add(tf.keras.layers.Dense(units=10, activation='softmax'))

### Compiling the model

- Optimizer: Adam
- Loss: Sparse softmax (categorical) crossentropy 

In [0]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['sparse_categorical_accuracy'])

In [0]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


### Training the model

In [0]:
model.fit(X_train, y_train, epochs=5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 5s 91us/sample - loss: 0.5297 - sparse_categorical_accuracy: 0.8128
Epoch 2/5
60000/60000 [==============================] - 4s 67us/sample - loss: 0.3970 - sparse_categorical_accuracy: 0.8547
Epoch 3/5
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3641 - sparse_categorical_accuracy: 0.8662
Epoch 4/5
60000/60000 [==============================] - 4s 66us/sample - loss: 0.3440 - sparse_categorical_accuracy: 0.8736
Epoch 5/5
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3293 - sparse_categorical_accuracy: 0.8792


### Model evaluation and prediction

In [0]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)

10000/10000 [==============================] - 1s 65us/sample - loss: 0.3606 - sparse_categorical_accuracy: 0.8700


In [0]:
print("Test loss: {}".format(test_loss))
print("Test accuracy: {}".format(test_accuracy))

Test loss: 0.3605656342029572
Test accuracy: 0.8700000047683716


From the above ANN, we see that our model makes the accuracy of 87% in test dataset which is approximately similar to the last epoch in training set without any parameter tunings, thanks to the dropout we used which prevent overfitting. The next step will be do to some parameter tuning and see if the model makes good predictions.